design/API_requirements_and_design.md

```diff
@@ -0,0 +1,59 @@
+# Thin-Edge API requirements and design
+
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
+* The tedge_api crate defines only messages related to the runtime as Shutdown.
+* The implementation of the mechanisms defined by the tedge_api are provided by
+  a tedge_runtime crate.
+* The final executable is build as an assemblage of all these plugin crates and
+  a `main.rs` that defines which plugins have to be instantiated and then
+  launched.
+
+---------
+
+# Non-functional Requirements
+
+- Plugins can be included in the final executable without having them 'instantiated'
+  - The motivation is to be able to deliver a battery included executable, ready
+    to be used for a large set of use-cases without any Rust expertise.
+  - Notably a battery-included executable should include an MQTT connection
+    plugin and a JSON plugin for telemetry messages.
+- Plugins are instantiated and configured through an external configuration at
+  runtime
+- At startup the configuration is read and from it a set of plugins 'to be
+  instantiated' is derived
+- One should be able to deduplicate parts of the configuration (to prevent
+  repetition)
```

Comment:
I'm not expecting something fancy here - just a way to share values between instances of the same plugin.

Looking at this [configuration example](https://github.com/matthiasbeyer/thin-edge.io/blob/feature/add_tedge_api_impl_typed/tedge/example-config.toml),
I wonder if this can be simply done by adding a level for all the instances of a plugin.

For instance:
```
[plugins.mqtt.configuration]
host = "127.0.0.1"
port = 1883
clear_session = false

[plugins.mqtt.c8y.configuration]
session = "c8y-mapper"
subscriptions = "c8y/#"

[plugins.mqtt.tedge.configuration]
session = "tedge"
subscriptions = "tedge/measurements#"
```